R/chop.R

```diff
@@ -90,188 +90,253 @@ chop <- function(data, cols) {
+    row_recycle <- col_sizes != sizes
+    col[row_recycle] <- map2(col[row_recycle], sizes[row_recycle], vec_recycle)
 
-    cols[[i]] <- col
-  }
+    col <- vec_unchop(col, ptype = col_ptype, name_spec = zap())
 
-  out_size <- sum(sizes)
+    if (is_null(col)) {
+      # This can happen when both of these are true:
+      # - `col` was an empty list(), or a list of all `NULL`s.
+      # - No ptype was specified for `col`, either by the user or by a list-of.
+      if (out_size != 0L) {
+        abort("Internal error: `NULL` column generated, but output size is not `0`.")
+      }
 
-  loc <- rep(seq_len_size, sizes)
+      col <- unspecified(0L)
+    }
 
-  val <- new_data_frame(cols, n = out_size)
-  if (!is.null(ptype)) {
-    val <- vec_cast(val, ptype)
+    out_cols[[i]] <- col
   }
 
-  out <- list(loc = loc, val = val)
+  names(out_cols) <- names
+  out_val <- new_data_frame(out_cols, n = out_size)
+
+  out <- list(loc = out_loc, val = out_val)
   out <- new_data_frame(out, n = out_size)
 
   out
 }
 
-df_unchop_ptype <- function(x) {
-  if (vec_is_list(x)) {
-    attr(x, "ptype") %||% unspecified(0L)
-  } else {
-    vec_ptype(x)
+unchop_col_info <- function(col, size, keep_empty) {
+  sizes <- list_sizes(col)
+  null <- vec_equal_na(col)
+
+  ptype <- attr(col, "ptype", exact = TRUE)
+
+  if (any(null)) {
+    # Always replace `NULL` elements with size 1 missing equivalent for recycling.
+    # These will be reset to `NULL` in `unchop_finalize()` if the
+    # entire row was missing and `keep_empty = FALSE`.
+
+    if (is_null(ptype)) {
+      replacement <- list(unspecified(1L))
+    } else {
+      replacement <- list(vec_init(ptype, n = 1L))
+      replacement <- new_list_of(replacement, ptype = ptype)
+    }
+
+    col <- vec_assign(col, null, replacement)
+    sizes[null] <- 1L
   }
-}
 
-tidyr_size_finalise <- function(size) {
-  if (is.na(size)) {
-    0L
-  } else {
-    size
+  if (keep_empty) {
+    # Remember, `NULL` elements are already handled above, so `sizes == 0L`
+    # will now only happen with typed empty elements.
+    empty_typed <- sizes == 0L
+
+    if (any(empty_typed)) {
+      # Replace empty typed elements with their size 1 equivalent
+
+      if (is_null(ptype)) {
+        # `vec_init()` is slow, see r-lib/vctrs#1423, so use `vec_slice()` equivalent
+        replacement <- map(vec_slice(col, empty_typed), vec_slice, i = NA_integer_)
+      } else {
+        # For list-of, all size elements are the same type
+        replacement <- list(vec_init(ptype, n = 1L))
+        replacement <- new_list_of(replacement, ptype = ptype)
+      }
+
+      col <- vec_assign(col, empty_typed, replacement)
+      sizes[empty_typed] <- 1L
+    }
   }
+
+  list(col = col, sizes = sizes, null = null)
 }
 
-tidyr_recycle <- function(x, size) {
-  if (is.null(x)) {
-    unspecified(size)
-  } else {
-    vec_recycle(x, size)
+unchop_sizes2 <- function(x, y) {
+  # Standard tidyverse recycling rules, just vectorized.
```

Comment:
Something that is different between this implementation and the previous one (and @mgirlich's) is that we now use standard recycling rules when computing the majority of the common sizes for the rows. No more awkward special handling of `NA_integer_` sizes!

To be able to do this, we do a pre-adjustment to `NULL` elements in `unchop_col_info()`, and then a post-adjustment to entirely `NULL` rows in `unchop_finalize()`